slatedb/src/wal_buffer.rs

```diff
@@ -0,0 +1,602 @@
+    task::JoinHandle,
+};
+
+use crate::{
+    db_state::SsTableId,
+    iter::KeyValueIterator,
+    mem_table::KVTable,
+    tablestore::TableStore,
+    types::RowEntry,
+    utils::{
+        spawn_bg_task, MonotonicClock, MonotonicSeq, WatchableOnceCell, WatchableOnceCellReader,
+    },
+    wal_id::WalIdStore,
+    SlateDBError,
+};
+
+/// [`WalBufferManager`] buffers write operations in memory before flushing them to persistent storage.
+/// The flush operation only targets Remote storage right now, later we can add an option to flush to local
+/// storage.
+///
+/// It maintains a `current_wal` buffer for active writes and a queue of immutable WALs pending flush.
+///
+/// By default, it offers a best-effort durability guarantee based on:
+///
+/// - `max_wal_size`: Flushes when `max_wal_size` bytes is exceeded
+/// - `max_flush_interval`: Flushes after `max_flush_interval` elapses, if set
+///
+/// For strict durability requirements on synchronous writes, use [`WalBufferManager::flush()`] to explicitly
+/// trigger a flush operation and await the result. This will flush ALL the in memory WALs (including the
+/// current WAL) to remote storage.
+///
+/// The manager is thread-safe and can be safely shared across multiple threads.
+///
+/// Please note:
+///
+/// - the size limit (`max_wal_size`) is a soft threshold. WAL entries within a single write batch are
+///   guaranteed to be written atomically to the same WAL file.
+/// - Fatal errors during flush operations are stored internally and propagated to all subsequent
+///   operations. The manager becomes unusable after encountering a fatal error.
+pub struct WalBufferManager {
+    inner: Arc<parking_lot::RwLock<WalBufferManagerInner>>,
+    wal_id_incrementor: Arc<dyn WalIdStore + Send + Sync>,
+    quit_once: WatchableOnceCell<Result<(), SlateDBError>>,
+    table_store: Arc<TableStore>,
+    mono_clock: Arc<MonotonicClock>,
+    max_wal_bytes_size: usize,
+    max_flush_interval: Option<Duration>,
+}
+
+struct WalBufferManagerInner {
+    current_wal: Arc<KVTable>,
+    /// When the current WAL is ready to be flushed, it'll be moved to the `immutable_wals`.
+    /// The flusher will try flush all the immutable wals to remote storage.
+    immutable_wals: VecDeque<(u64, Arc<KVTable>)>,
+    /// The channel to send the flush work to the background worker.
+    flush_tx: Option<mpsc::Sender<WalFlushWork>>,
+    /// task handle of the background worker.
+    background_task: Option<JoinHandle<Result<(), SlateDBError>>>,
+    /// Whenever a WAL is applied to Memtable and successfully flushed to remote storage,
+    /// the immutable wal can be recycled in memory.
+    last_applied_seq: Option<u64>,
+    /// The flusher will update the recent_flushed_wal_id and last_flushed_seq when the flush is done.
+    recent_flushed_wal_id: u64,
+    /// The last flushed sequence number.
+    last_remote_persisted_seq: Arc<MonotonicSeq>,
+}
+
+impl WalBufferManager {
+    pub fn new(
```

Comment:
Can all WalBufferManager be `pub(crate)` instead of `pub`?

Reply:
fixed in c2aa219